File: graphs/bidirectional.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Edge
{
public:
    int v1;
    int v2;
    int wt;
    Edge(int v1, int v2, int wt)
    {
        this->v1 = v1;
        this->v2 = v2;
        this->wt = wt;
    }
};
class Mypair
{
public:
    int wsf = 0;
    string psf = "";
    Mypair(int wsf, string psf)
    {
        this->wsf = wsf;
        this->psf = psf;
    }
};
class MyEpair
{
public:
    int wsf = 0;
    string psf = "";
    int noOfE = 0;
    MyEpair(int wsf, string psf, int noOfE)
    {
        this->wsf = wsf;
        this->psf = psf;
        this->noOfE = noOfE;
    }
};

vector<vector<Edge *>> graph;
//getIdx
int getIdx(int v1, int v2)
{
    for (int i = 0; i < graph[v1].size(); i++)
    {
        if (graph[v1][i]->v2 == v2)
        {
            return i;
        }
    }
    return -1;
}
//addVertex
void addVertex(int v1)
{
    if (graph.size() == v1)
    {
        vector<Edge *> innergraph;
        graph.push_back(innergraph);
    }
}
//rmVertex
void removeVertex(int v1)
{
    for (Edge *e : graph[v1])
    {
        int idx = getIdx(e->v2, v1);
        graph[e->v2].erase(graph[e->v2].begin() + idx);
    }
    graph.erase(graph.begin() + v1);
}

//addEdges
void addEdges(int v1, int v2, int wt)
{
    if (v1 < graph.size() && v2 < graph.size())
    {

        graph[v1].push_back(new Edge(v1, v2, wt));
        graph[v2].push_back(new Edge(v2, v1, wt));
    }
}
//rmEdge
void removeEdge(int v1, int v2)
{
    int idx1 = getIdx(v1, v2);
    if (idx1 != -1)
    {
        graph[v1].erase(graph[v1].begin() + idx1);
    }
    int idx2 = getIdx(v2, v1);
    if (idx2 != -1)
    {
        graph[v2].erase(graph[v2].begin() + idx2);
    }
}

void display()
{
    string sb = "";
    for (int i = 0; i < graph.size(); i++)
    {
        for (Edge *e : graph[i])
        {
            sb += "(" + to_string(e->v1) + "," + to_string(e->v2) + ")" + " @" + to_string(e->wt) + " ";
        }
        sb += "\n";
    }
    cout << sb << endl;
}
void solve()
{
    for (int i = 0; i < 7; i++)
    {
        addVertex(i);
    }

    addEdges(0, 1, 10);
    addEdges(0, 3, 40);
    addEdges(1, 2, 10);
    addEdges(2, 3, 10);
    addEdges(3, 4, 2);
    addEdges(4, 5, 2);
    addEdges(4, 6, 8);
    addEdges(5, 6, 3);

    display();
    vector<bool> isVisited([graph.size()], false);
    Mypair *p = minwtPath(0, 6, isVisited);
    cout << p->wsf << p->psf;
}

//sw
Mypair *minwtPath(int src, int desti, vector<bool> isVisited)
{
    if (src == desti)
    {
        Mypair *p = new Mypair(0, to_string(src));
        return p;
    }
    Mypair *myAns = new Mypair(10000, "");

    isVisited[src] = true;
    for (Edge *e : graph[src])
    {
        if (isVisited[e->v2])
            continue;

        Mypair *recAns = minwtPath(e->v2, desti, isVisited);
        if (recAns->wsf + e->wt < myAns->wsf)
        {
            myAns->wsf = recAns->wsf + e->wt;
            myAns->psf = to_string(src) + recAns->psf;
        }
    }

    isVisited[src] = false;
    return myAns;
}
//sp
MyEpair *shortestPath(int src, int desti, vector<bool> isVisited)
{
    if (src == desti)
    {
        MyEpair *p = new MyEpair(0, to_string(src), 0);
        return p;
    }
    MyEpair *myAns = new MyEpair(0, "", 1000000);

    isVisited[src] = true;
    for (Edge *e : graph[src])
    {
        if (isVisited[e->v2])
            continue;

        MyEpair *recAns = shortestPath(e->v2, desti, isVisited);
        if (recAns->noOfE + 1 < myAns->noOfE)
        {
            myAns->wsf = recAns->wsf + e->wt;
            myAns->psf = to_string(src) + recAns->psf;
            myAns->noOfE = recAns->noOfE + 1;
        }
    }

    isVisited[src] = false;
    return myAns;
}

int main()
{

    solve();

    return 0;
}

File: graphs/bidirectional.java
import java.util.*;

public class bidirectional {

    public static void main(String args[]) {
        solve();

    }

    public static void solve() {
        for (int i = 0; i < 7; i++) {
            addVertex(i);
        }
        addEdges(0, 1, 10);
        addEdges(0, 3, 40);
        addEdges(1, 2, 10);
        addEdges(2, 3, 10);
        addEdges(3, 4, 2);
        addEdges(4, 5, 2);
        addEdges(4, 6, 8);
        addEdges(5, 6, 3);

        display();
        System.out.println(noOfEdges);
        System.out.println(totalNoOfEdges());
        removeEdge(3, 4);
        removeVertex(3);
    }

    public static class Edge {

        int v1 = 0;
        int v2 = 0;
        int weight = 0;

        public Edge(int v1, int v2, int weight) {
            this.v1 = v1;
            this.v2 = v2;
            this.weight = weight;
        }
    }

    private static ArrayList<ArrayList<Edge>> graph = new ArrayList<>();
    private static int noOfEdges = 0;

    // addVertex
    public static void addVertex(int v1) {
        if (graph.size() == v1) {
            graph.add(new ArrayList<>());
        }
    }

    // addEdges
    public static void addEdges(int v1, int v2, int weight) {
        if (v1 < graph.size() && v2 < graph.size()) {
            graph.get(v1).add(new Edge(v1, v2, weight));
            graph.get(v2).add(new Edge(v2, v1, weight));
            noOfEdges++;
        }
    }

    // total no of edges
    public static int totalNoOfEdges() {
        int count = 0;
        for (int i = 0; i < graph.size(); i++) {
            count += graph.get(i).size();
        }
        return count / 2;
    }

    // remove edges
    public static void removeEdge(int v1, int v2) {
        int idx1 = getIdx(v1, v2);
        if (idx1 != -1) {
            graph.get(v1).remove(idx1);
        }
        int idx2 = getIdx(v2, v1);
        if (idx2 != -1) {
            graph.get(v2).remove(idx2);
        }
    }
    //part of remove edges
    public static int getIdx(int v1, int v2) {
        for (int i = 0; i < graph.get(v1).size(); i++) {
            Edge e = graph.get(v1).get(i);
            if (e.v2 == v2) {
                return i;
            }
        }
        return -1;
    }

    // remove vertex
    public static void removeVertex(int v1) {
        for (int i = 0; i < graph.get(v1).size(); i++) {
            Edge e = graph.get(v1).get(i);
            int idx = getIdx(e.v2, e.v1);
            graph.get(e.v2).remove(idx);
        }
        graph.get(v1).remove(v1);
    }

    // contains
    public static boolean contains(int v1, int v2) {
        if (v1 < graph.size() && v2 < graph.size() && getIdx(v1, v2) != -1) {
            return true;
        }
        return false;
    }

    // display
    public static void display() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < graph.size(); i++) {
            for (Edge e : graph.get(i)) {
                sb.append("(" + e.v1 + "," + e.v2 + ")" + " @ " + e.weight);
            }
            sb.append("\n");
        }
        System.out.print(sb.toString());
    }

    // ATOB path
    public static void hasPath() {
        boolean[] isVisited = new boolean[graph.size()];
        System.out.println(hasPath(0, 6, isVisited, ""));
    }

    public static boolean hasPath(int src, int desti, boolean[] isVisited, String ans) {
        if (src == desti) {
            System.out.print(res + " - >");
            return true;
        }
        if (isVisited[src]) {
            for (Edge e : graph.get(src)) {
                if (isVisited[e.v2])
                    continue;
                boolean res = hasPath(e.v2, desti, isVisited, ans + src);

                if (res) {
                    return true;
                }
            }

        }
        return false;
    }
    // ATOB ALL paths
    // ATOB SHORTEST paths

}